worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,326 @@
+	filter         func(changestream.ChangeEvent) bool
+}
+
+// EventQueue defines an event listener and dispatcher for db changes that can
+// be multiplexed to subscriptions. The event queue allows consumers to
+// subscribe via callbacks to the event queue. This is a lockless
+// implementation, all subscriptions and changes are serialized in the main
+// loop. Dispatching is randomized to ensure that subscriptions don't depend on
+// ordering. The subscriptions can be associated with different subscription
+// options, which provide filtering when dispatching. Unsubscribing is provided
+// per subscription, which is done asynchronously.
+type EventQueue struct {
+	catacomb catacomb.Catacomb
+	stream   Stream
+	logger   Logger
+
+	subscriptions      map[uint64]*subscription
+	subscriptionsByNS  map[string][]*eventFilter
+	subscriptionsAll   map[uint64]struct{}
+	subscriptionsCount uint64
+
+	// (un)subscription related channels to serialize adding and removing
+	// subscriptions. This allows the queue to be lock less.
+	subscriptionCh   chan subscriptionOpts
+	unsubscriptionCh chan uint64
+}
+
+// New creates a new EventQueue that will use the Stream for events.
+func New(stream Stream, logger Logger) (*EventQueue, error) {
+	queue := &EventQueue{
+		stream:             stream,
+		logger:             logger,
+		subscriptions:      make(map[uint64]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsAll:   make(map[uint64]struct{}),
+		subscriptionsCount: 0,
+
+		subscriptionCh:   make(chan subscriptionOpts),
+		unsubscriptionCh: make(chan uint64),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &queue.catacomb,
+		Work: queue.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return queue, nil
+}
+
+// Subscribe creates a new subscription to the event queue. Options can be
+// provided to allow filter during the dispatching phase.
+func (q *EventQueue) Subscribe(handler changestream.Handler, opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	// Get a new subscription count without using any mutexes.
+	subID := atomic.AddUint64(&q.subscriptionsCount, 1)
+
+	sub := newSubscription(subID, handler, func() { q.unsubscribe(subID) })
+	if err := q.catacomb.Add(sub); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		select {
+		case <-q.catacomb.Dying():
+			return
+		case q.subscriptionCh <- subscriptionOpts{
+			subscription: sub,
+			opts:         opts,
+		}:
+		}
+	}()
+
+	select {
+	case <-done:
+	case <-q.catacomb.Dying():
+		return nil, q.catacomb.ErrDying()
+	}
+
+	return sub, nil
+}
+
+// Kill stops the event queue.
+func (q *EventQueue) Kill() {
+	q.catacomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (q *EventQueue) Wait() error {
+	return q.catacomb.Wait()
+}
+
+func (q *EventQueue) unsubscribe(subscriptionID uint64) {
+	go func() {
```

Comment:
Just make this synchronise, I don't understand why you want unsubscribe to not offer the guarantees that a callback function won't be called after the unsubscribe returns. It makes cleanup code a simple `defer sub.Unsubscribe()`. Callback functions are dangerous without clear, well defined hard guarantees on when they will be called, because people make mistakes.